vowpalwabbit/csoaa.cc

```diff
@@ -398,51 +400,60 @@ void do_actual_learning_oaa(ldf& data, base_learner& base, size_t start_K)
   }
 }
 
+
+/*
+* The begining of the multi_ex sequence may be labels.  Process those
+* and return the start index of the un-processed examples
+*/
+multi_ex process_labels(ldf& data, const multi_ex& ec_seq_all);
+
+/*
+ * 1) process all labels at first 
+ * 2) verify no labels in the middle of data
+ * 3) learn_or_predict(data) with rest
+ */
 template <bool is_learn>
-void do_actual_learning(ldf& data, base_learner& base)
+void do_actual_learning(ldf& data, single_learner& base, multi_ex& ec_seq_all)
 {
-  if (data.ec_seq.size() <= 0) return;  // nothing to do
-  /////////////////////// handle label definitions
+  if (ec_seq_all.size() == 0) return;  // nothing to do
 
-  if (ec_seq_is_label_definition(data.ec_seq))
-  {
-    for (size_t i=0; i<data.ec_seq.size(); i++)
-    {
-      features new_fs = data.ec_seq[i]->feature_space[data.ec_seq[i]->indices[0]];
+  data.ft_offset = ec_seq_all.last()->ft_offset;
```

Comment:
The first element is canonical for the prediction, so maybe we should make that the rule?

Reply:
Todo: shift to first.

Reply:
Done